src/core/NewRestriction.ml

```diff
@@ -0,0 +1,213 @@
+  match x with
+  | `Atom x -> let x, h = reserve_atom x h in `Atom x, h
+  | `Dim0 -> `Dim0, h
+  | `Dim1 -> `Dim1, h
+
+let union (x : dim) (y : dim) h =
+  let x, h = reserve x h in
+  let y, h = reserve y h in
+  union_cls x y h
+
+let query_atom (x : atom) (index : (atom, int) T.t) : [`Ok of cls | `Owned of atom] =
+  try
+    `Ok (`Atom (T.get x index))
+  with
+    _ ->
+    `Owned x
+
+let query (x : dim) (index : (atom, int) T.t) : [`Ok of cls | `Owned of atom] =
+  match x with
+  | `Atom x -> query_atom x index
+  | `Dim0 -> `Ok `Dim0
+  | `Dim1 -> `Ok `Dim1
+
+let compare x y (h : t) =
+  match query x h.index, query y h.index with
+  | `Owned x, `Owned y ->
+    if x == y then `Same else `Indet
+  | `Owned _, `Ok _ -> `Indet
+  | `Ok _, `Owned _ -> `Indet
+  | `Ok x, `Ok y ->
+    match find_cls x h, find_cls y h with
+    | `Dim0, `Dim0 | `Dim1, `Dim1 -> `Same
+    | `Dim0, `Dim1 | `Dim1, `Dim0 -> `Apart
+    | `Atom x, `Atom y ->
+      if x == y then `Same else `Indet
+    | `Atom _, _ -> `Indet
+    | _, `Atom _ -> `Indet
+
+let mem_atom (x : atom) index =
+  T.mem x index
+
+let hide_aux (x : atom) (h : t) =
+  {h with index = T.remove x h.index}
+
+let hide (x : atom) (h : t) =
+  if mem_atom x h.index then
+    `Changed (hide_aux x h)
+  else
+    `Same
+
+let hide' (x : atom) (h : t) =
+  if mem_atom x h.index then
+    hide_aux x h
+  else
+    h
+
+let equate = union
+
+let equate' x y h =
+  get_m h @@ equate x y h
+
+let subst (r : dim) (x : atom) (h : t) =
+  if mem_atom x h.index && r <> `Atom x then
+    `Changed (hide_aux x (equate' (`Atom x) r h))
+  else
+    `Same
+
+let subst' r x h =
+  if mem_atom x h.index && r <> `Atom x then
+    hide_aux x (equate' (`Atom x) r h)
+  else
+    h
+
+let swap (x : atom) (y : atom) (h : t) =
+  match T.find x h.index, T.find y h.index with
+  | None, None -> h
+  | Some idx, Some idy when idx = idy -> h
+  | oidx, oidy -> {h with index = T.set_opt y oidx (T.set_opt x oidy h.index)}
+
+
+let pp_cls fmt =
+  function
+  | `Dim0 ->
+    Format.fprintf fmt "0"
+  | `Dim1 ->
+    Format.fprintf fmt "1"
+  | `Atom i ->
+    Format.fprintf fmt "#%a" Format.pp_print_int i
+
+let pp_parent fmt (x, cls) =
+  Format.fprintf fmt "%a~%a" Name.pp x pp_cls cls
+
+let pp fmt h =
+  let comma fmt () = Format.fprintf fmt "," in
+  let l = T.fold (fun x i l -> (x, find_index i h) :: l) h.index [] in
+  Format.pp_print_list ~pp_sep:comma pp_parent fmt l
+
+(* poor man's tests *)
+let _  =
+  try
+    let x = `Atom (Name.named (Some "i")) in
+    let rst = equate' x `Dim1 @@ emp () in
+    let rst = equate' x `Dim0 rst in
+    failwith "Test failed"
+  with
+  | I.Inconsistent -> ()
+
+let _ =
+  let x = `Atom (Name.named (Some "i")) in
+  let rst = equate' x `Dim0 @@ emp () in
+  assert (compare x `Dim0 rst == `Same)
+
+
+
+
+
+let from_old_restriction rst =
+  let eqns = Restriction.chronicle rst in
+  List.fold_left (fun rel (r, r') -> equate' r r' rel) (emp ()) eqns
```

Comment:
Looks okay to me now. Hope this will be killed by the New Order eventually.

Reply:
yes! Once we can delete all the old stuff, we can delete this too.